### PR TITLE
simplified event handling logic by screening undesirable events

### PR DESCRIPTION
**Pull request by zumaad**
We get a bunch of duplicate events so I made a method to reject them. Also, when a offer finishes buying or selling, we get a BUYING/SELLING event followed by a BOUGHT/SOLD event. Since only one of them is needed, the method only lets the BOUGHT/SOLD event go through. 

Essentially, all undesirable events cause the method to return true. The method is called in 
```onGrandExchangeOfferChanged``` and is passed the ```newOfferEvent``` which it then used to decide whether to return true or false. If true is returned,```onGrandExchangeOfferChanged``` will return, hence that event doesn't get processed. 

The reason I added this is because we are going to be working on the profit tracker and soon we will have to process every type of event and not just the ones that are relevant to margin checking/subtracting from GE limits. So, this method will make that a lot easier as we don't have to worry about duplicate events and such. Here are some logs to show the difference.

BEFORE: 
Screen Shot 2020-04-04 at 1 13 30 AM

Notice all the duplicate BUYING and SELLING events and whenever a buy/sell is completed, you get both a BUYING/SELLING event and a BOUGHT/SOLD event.

This is after:

Screen Shot 2020-04-04 at 2 46 26 AM

No duplicate events, every event conveys a unique piece of information :)

### CODE REVIEW

**Comment by Belieal** [comment]
Hmm, I wonder if this is the same thing that happens to the widget as onVarClientInt gets called twice every single time. I'll have to test if your way of checking for duplicates also works there. Should probably also open an issue on the pluginhub repo, since I'm pretty sure this behavior is present purely in events inside external plugins.

Before I merge it, would you mind running it through with the RuneLite checkstyle? 😄

**Comment by zumaad** [comment]
I ran it through the Runelite check style :) It shows a ton of lines changed now tho, but it does look better haha.

**Comment by Belieal** [comment]
So I seem to have found out why we got duplicate events. It just so happens that if you have the plugin installed from the plugin hub and then open the plugin through the test main in IntelliJ, it'll think you have two of the exact same plugins running and somehow it thinks it's the same instance? It's really weird, but regardless this needs to be changed/reverted else cancelled buy offers won't update the remaining GE limit in release.

**Comment by zumaad** [comment]
wow that is really weird. So when you open it regularly through runelite you don't get duplicate events? However, even if we don't get duplicate events this will let the events pass through as it checks if its equal (state and quantity) to the last offer for that slot, which, if we don't get duplicate events, wouldn't happen.

File: src/main/java/com/flippingutilities/FlippingPlugin.java [hunk 1]
```diff
@@ -1,609 +1,658 @@
...
+		//May change this in the future to be able to update prices independent of quantity.
+		//Perhaps some way of timing the state change from buying to bought such that we know it was "instantly" bought/sold?
+		if ((newOfferState == GrandExchangeOfferState.BOUGHT
+			|| newOfferState == GrandExchangeOfferState.SOLD) && newOffer.getQuantitySold() == 1)
+		{
+			addFlipTrade(tradeConstructor(newOffer));
+			panel.rebuildFlippingPanel(tradesList);
+		}
+		//If the new offer is of state BOUGHT Record the trade to keep track of GE limit.
+		else if (newOffer.getQuantitySold() > 0 && newOfferState == GrandExchangeOfferState.BOUGHT)
```
Comment (Belieal) [inline src/main/java/com/flippingutilities/FlippingPlugin.java:311]: I think this change might've been the cause for the loss of cancelled buy trades going through. Any objections to amend the condition with an ``` || newOfferState == GrandExchangeOfferState.CANCELLED_BUY)```?

Comment (zumaad) [reply src/main/java/com/flippingutilities/FlippingPlugin.java:311]: ahhh yes that makes sense. Good catch! Yeah that change should be good. I've been working on some code that delegates the responsibility of managing trade limits and refresh times to the history manager, its nice and concise due to access to standardized trades and it fixes a potential bug. Just gonna write some comments and submit a PR soon.

**Pull request merged**